### ISSUE TITLE
Building on freebsd with clang++

### ISSUE THREAD

**Issue by josser** [Sembiance/mhash#14]
Hi, found strange issue with mhash on freebsd. 

First of all I know that there is a node-gyp issue, described here https://github.com/Sembiance/mhash/issues/11 and I have created symlink to clang++: 

```
ls -la /usr/bin/g++
lrwxr-xr-x  1 root  wheel  16 19 feb 09:56 /usr/bin/g++ -> /usr/bin/clang++
```

But now, I've got another issue, here is a build-log: https://gist.github.com/josser/d3e4524be46afccc5227 

I'm not familiar with c++ so if there are any things I missed, feel free to ask.
Thanks, and sorry for my bad english :)

**Comment by Sembiance** [comment]
Hello

I took a look at the gist and don't know what exactly might be wrong. It's clearly a bug in the way the mhash lib is being compiled but I don't know what exactly is wrong without further digging.

So I downloaded the FreeBSD 10.1 amd64 virtual disk image and I was going to test this out. Sadly I couldn't even get node to compile and install, even after symlinking g++ to clang++.

**Comment by josser** [comment]
Use ports! ) 

```bash
cd /usr/ports/www/node
make install clean 
```

**Comment by josser** [comment]
Ok, I have some new information about this issue. 

The error goes from ../deps/mhash/include/mutils/mhash.h:118:1: error: unknown type name '_Bool'
You may found this link useful: https://llvm.org/bugs/show_bug.cgi?id=21410 

This is actual part from the file which lead to an error: 
```c++
/*
 * Due to buggy implementations of the boolean headers in some systems,
 * we have to detect the boolean values seperately from the boolean type.
 * A big thank you to Heiko Lehmann for spotting the bug. Unfortunately,
 * this fix may not be enough, as it is only going to work on things the
 * precompiler knows about.
 *
 * Why go to all this trouble over something that should have a standard
 * value anyway? Because.
 */
#if defined(HAVE__BOOL)
#define mutils_boolean _Bool
#else
typedef char mutils_boolean;
#endif
```
Than I went to ports tree and check, how guys from freebsd compiles this library. And here we go: 
http://svnweb.freebsd.org/ports/head/security/mhash/files/patch-include-mutils-mutils.h?view=markup

This is patch which they apply on top of the mhash source before building it for freebsd. 
As you can see, there is nothing about mutils_boolean, so I downloaded source and tried to build it myself. 

Mystically it builds fine without any errors.

So here I'm stuck because my knowledge of C/C++ is to weak.

The only thing I understand is that issue is in _Bool type and some difference in C/C++: 
Here is example: 
```c
#include <stdbool.h>

int main(){
    _Bool y;
    return 0;
}
```
If I do:
$clang test.c 
compiles fine, but compiling it as c++:  
$clang test.cc 
gives compilation error: test.cc:4:5: error: unknown type name '_Bool'


Hope this can help and sorry for my bad english again :)

**Comment by Sembiance** [comment]
Maybe I should have node-mhash only compile mhash if not found on the system. This way folks can install mhash separately and node-mhash will just use the system lib.

**Comment by josser** [comment]
I have found a way to build it with this hack: 

```c++
#define __GNUC__ 4

#include <stdbool.h>

int main(){
    _Bool y;
    return 0;
}
```

$clang test.cpp 
works without any errors. Maybe you can add this but I don't know possible drawbacks

**Comment by Grumbel** [comment]
I am running into the same issue with `_Bool` on Ubuntu. With g++ it works fine, but with clang++ it fails since `_Bool` isn't defined. As far as I can tell the issue is that clang does only define `_Bool` in C mode or in GNUC mode, but not when you use C++:

http://clang.llvm.org/doxygen/stdbool_8h_source.html

The `mhash_config.h` file can't catch the issue as both g++ and clang++ use the file provided by the precompiled package, which is probably compiled with gcc.

The proper fix as I understand it would be to use `bool` instead of `_Bool`, but the comment in `mutils/mutils.h` suggest that that might cause trouble elsewhere. A brute force fix would be to just check for a `_Bool` define:

```
#ifdef _Bool
#  define mutils_boolean _Bool
#else
#  define mutils_boolean bool
#endif
```

**Comment by miguelaferreira** [comment]
Did anyone find a workaround this issues?

I'm having the same kind of trouble building on alpine:
```
...
make[1]: Leaving directory '/code/node_modules/mhash/deps/mhash'
  TOUCH Release/obj.target/libmhash.stamp
Release/obj.target/libmhash.stamp Stats {
  dev: 44,
  mode: 33188,
  nlink: 1,
  uid: 0,
  gid: 0,
  rdev: 0,
  blksize: 4096,
  ino: 13262690,
  size: 0,
  blocks: 0,
  atime: 2017-07-07T08:04:51.000Z,
  mtime: 2017-07-07T08:04:51.000Z,
  ctime: 2017-07-07T08:04:51.000Z,
  birthtime: 2017-07-07T08:04:51.000Z }
  CXX(target) Release/obj.target/mhash/mhash.o
In file included from ../deps/mhash/include/mutils/mtypes.h:24:0,
                 from ../deps/mhash/include/mhash.h:8,
                 from ../mhash.cc:7:
../deps/mhash/include/mutils/mutils.h:100:24: error: '_Bool' has not been declared
 #define mutils_boolean _Bool
                        ^
../deps/mhash/include/mutils/mutils.h:250:70: note: in expansion of macro 'mutils_boolean'
 mutils_word32 *mutils_word32nswap(mutils_word32 *x, mutils_word32 n, mutils_boolean destructive);
...
```

node -v v6.10.3
node-gyp -v v3.4.0

**Comment by miguelaferreira** [comment]
I've submitted a PR that fixed my build #20

**Comment by Sembiance** [comment]
Should be fixed with pull request #20 having been merged.
